src/audio/module_adapter/module_adapter.c

```diff
@@ -241,20 +241,31 @@ int module_adapter_prepare(struct comp_dev *dev)
 	}
 
 	/* allocate memory for input buffers */
-	mod->input_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
-				     sizeof(*mod->input_buffers) * mod->num_input_buffers);
-	if (!mod->input_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
-		return -ENOMEM;
+	if (mod->num_input_buffers) {
+		mod->input_buffers =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*mod->input_buffers) * mod->num_input_buffers);
+		if (!mod->input_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
+			return -ENOMEM;
+		}
+	} else {
+		mod->input_buffers = NULL;
 	}
 
 	/* allocate memory for output buffers */
-	mod->output_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
-				      sizeof(*mod->output_buffers) * mod->num_output_buffers);
-	if (!mod->output_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
-		ret = -ENOMEM;
-		goto in_free;
+	if (mod->num_output_buffers) {
+		mod->output_buffers =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*mod->output_buffers) * mod->num_output_buffers);
+		if (!mod->output_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
+			if (mod->input_buffers)
+				rfree(mod->input_buffers);
+			return -ENOMEM;
```

Comment:
`rfree(NULL)` is ok. Please use `goto` and add a label above line 410 below

Reply:
@lyakh But isn't it easier here to just `rfree(mod->input_buffers)` even without a NULL check and return an error? At that  point only `mod->input_buffers` could be allocated and there is no need to free `mod->output_buffers` and jump to label.

Reply:
I'm ok with both. Traditionally, if you have a goto-based rollback of resources, you can return immediately on first error when there is no resources to free yet (so goto is not needed), but once you an hit error, you should use the rollbacks and jump to to a goto label. Of course, if you don't have have a rollback and you just free resources at location of the error and return, that's different, but you should not be mixing approaches. Either you have a rollback and you use it, or you don't. But as said, this is minor stuff, not blocking PR because of this.

Reply:
@kv2019i Thanks for explanation. Add a label to be aligned with the "goto label" approach.